File: webserver/rlnwebportal/rlnwebportal/mqtt.py
import django
django.setup()
import paho.mqtt.client as mqtt
import json
from datetime import datetime
from main.models import entries, dim_node, dim_entry_status, dim_node_status

broker_ip = "192.168.0.22"
topic = "RLN/gateway"

def on_connect(client, userdata, rc, properties=None):
	client.subscribe(topic)

def on_message(client, userdata, msg):
	print(msg.payload.decode())
	n = dim_node
	es = dim_entry_status
	try:
		dt = datetime.now()
		json_payload = json.loads(msg.payload.decode())
		name = json_payload["n"]
		add = json_payload["add"]
		hpno = json_payload["hp"]
		count = json_payload["cnt"]
		danger = bool(json_payload["dg"])
		firstaid = bool(json_payload["fa"])
		water = bool(json_payload["wt"])
		food = bool(json_payload["fd"])
		hug = bool(json_payload["hg"])
		node = n.objects.get(pk=json_payload["node"])
		status = es.objects.get(pk=1)
		entry = entries(name = name, address = add, hpno = hpno, occupants = count, danger = danger, firstaid = firstaid, water = water, food = food, hug = hug, node = node, status = status, datetime = dt)
		entry.save()
	except json.JSONDecodeError:
		print(msg.payload)
	except TypeError:
		print(msg.payload)
	except KeyError:
		print(msg.payload)
	pass

client = mqtt.Client()
client.on_connect = on_connect
client.on_message = on_message

client.connect(broker_ip, 1883, 60)

File: webserver/rlnwebportal/static/js/mqtt.js
var mqtt;
var reconnectTimeout = 2000;
var host;
var port;
var topic = "RLN/gateway";
function MQTTsetParam(hosts, ports){
	host = hosts;
	port = ports;
}
function onFailure(message) {
	console.log("Connection Attempt to Host "+host+"Failed");
	setTimeout(MQTTconnect, reconnectTimeout);
}
function onMessageArrived(msg){	
	// out_msg = "Message received "+msg.payloadString+"<br>";
	// out_msg = out_msg+"Message received Topic "+msg.destinationName;
	item = JSON.parse(msg.payloadString);
	if(typeof item['update'] == 'undefined'){ //This condition does not contain 'update' key
		console.log(msg.payloadString);
		tfkey = ["dg", "hg", "wt", "fa"];
		for(i = 0; i<tfkey.length; i++){
			if(typeof item[tfkey[i]] == "undefined" || item[tfkey[i]] == 0){
				item[tfkey[i]] = false;
			}
			else{
				item[tfkey[i]] = true;
			}
		}
		$.ajax({
			url: "api/manageEntries",
			dataType:'json',
			data:{
				"latest_id": ""
			},
			success: function(data){     
				id = data['data'][0]['latest_id'];
				buildArray(item, id);
			}
		});
		makePending();
	}
	else{//This condition contain 'update' key
		if(msg.payloadString!=selfUpdateJSON){
			
			var findID = item["id"];
			res = findEntry(ongoing_arr, findID);
			if(typeof res == "undefined"){
				res = findEntry(completed_arr, findID);
			}
			res[0]["status"] = item["update"];
			if(res[0]["status"] == 3){
				ongoing_arr.splice(res[1], 1);
				completed_arr.push(res[0]);
			}
			else{
				completed_arr.splice(res[1], 1);
				ongoing_arr.push(res[0]);
			}
			makePending();
			// if(res[0]["status"] != 3){
			// 	ongoing_arr[index] = res;
			// }
			// else{
			// 	ongoing_arr.splice(index, 1);
			// 	completed_arr.push(res);
			// }
			// makePending();
		}
	}
	
}

function onConnect() {
	console.log("Connected ");
	mqtt.subscribe(topic);
	// message = new Paho.MQTT.Message("Hello World");
	// message.destinationName = topic;
	// mqtt.send(message);
}
function MQTTconnect() {
	console.log("connecting to "+ host +" "+ port);
	var x=Math.floor(Math.random() * 10000); 
	var cname="orderform-"+x;
	mqtt = new Paho.MQTT.Client("192.168.0.22",9001,cname);
	var options = {
		timeout: 3,
		onSuccess: onConnect,
		onFailure: onFailure,
		};
	mqtt.onMessageArrived = onMessageArrived

	mqtt.connect(options); //connect
}

function MQTTsend(msg){
	message = new Paho.MQTT.Message(msg);
	message.destinationName = topic;
	mqtt.send(message);
}

function buildArray(item, id){
	currentdate = new Date(); 
	var datetime =  currentdate.getDate() + "/"
					+ (currentdate.getMonth()+1)  + "/" 
					+ currentdate.getFullYear() + " "
					+ currentdate.getHours() + ":"  
					+ currentdate.getMinutes() + ":" 
					+ currentdate.getSeconds();
	var post = {
		id: id,
		name: item['n'],
		address: item['add'],
		datetime: datetime,
		status: "Incomplete",
		nodeid: JSON.stringify(nodelist[item['node']]),
		hpno: item['hp'],
		occupants: item['cnt'],
		danger: item['dg'],
		hug: item['hg'],
		water: item['wt'],
		firstaid: item['fa'],
	}
	ongoing_arr.unshift(post);
}

function findEntry(arr, findID){
	for (var i=0; i < arr.length; i++) {
		if (arr[i].id == findID) {
			return [arr[i],i];
		}
	}
}